### ISSUE TITLE
Looper not initialized in new S3 TransferUtility

### ISSUE THREAD

**Issue by neurolab** [aws/aws-sdk-android#51]
Hi there,

Overall the new S3 Transfer Utility seems great!
I'm currently trying to upgrade our application to make use of the new API, but it seems the threading was not correctly implemented. I'm getting the following error when I try to upload a file : 
`Can't create handler inside thread that has not called Looper.prepare()`

It comes from the Handler() creation here :
https://github.com/aws/aws-sdk-android/blob/master/aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/TransferObserver.java#L161

Stacktrace : 
```
android.os.Handler.<init>(Handler.java:197)
android.os.Handler.<init>(Handler.java:111)
com.amazonaws.mobileconnectors.s3.transferutility.TransferObserver$TransferContentObserver.<init>(TransferObserver.java:161)
com.amazonaws.mobileconnectors.s3.transferutility.TransferObserver.<init>(TransferObserver.java:66)
com.amazonaws.mobileconnectors.s3.transferutility.TransferUtility.upload(TransferUtility.java:161)
com.neurodesign.utils.amazonws.AWSBridge.asyncS3Upload(AWSBridge.java:158)
com.neurodesign.utils.amazonws.AWSBridge.s3Upload(AWSBridge.java:132)
com.neurodesign.utils.s3PhotoUploader.S3PhotoUploadJob.onRun(S3PhotoUploadJob.java:74)
com.path.android.jobqueue.Job.safeRun(Job.java:189)
com.path.android.jobqueue.JobHolder.safeRun(JobHolder.java:91)
com.path.android.jobqueue.executor.JobConsumerExecutor$JobConsumer.run(JobConsumerExecutor.java:306)
java.lang.Thread.run(Thread.java:856)
```

**Comment by fosterzhang** [comment]
Thanks for reporting the issue. We are currently working on it. To unblock you temporarily, please call transfer utility on the main thread.

**Comment by corbt** [comment]
Hey, I'm running into this same issue with the TransferUtility. I need to be able to run my transfers from a background thread (all the synchronization goes on in an Android SyncAdapter) and I get this same error. When I try to manually call Looper.prepare() and then Looper.loop() in my background service, the transfer still seems to be stuck in the `TransferState.WAITING` state and nothing gets sent to s3.

Is there any roadmap/timeline for when this functionality will be available? Thanks.

**Comment by fosterzhang** [comment]
@corbt Do you mind sharing your code snippet of how you plan to use TransferUtility? We have the fix and plan to ship it in the coming release, but it'd be nice to verify it with your sample code.

**Comment by corbt** [comment]
Unfortunately the full code for a [SyncAdapter implementation](https://developer.android.com/training/sync-adapters/creating-sync-adapter.html) requires quite a number of files and configurations so the sync is performed properly in a background service, so it's more than would make sense to put here.

However, as long as it works in a background thread it should work for my use case. I've attached a simple test case below.

Also, it would be awesome if this library contained a method analogous to the TransferManager Upload's `waitForCompletion()` blocking method. I have a bunch of files I need to upload as part of the sync process but I'd rather take care of them serially (I don't want users trying to upload 20 files at once and killing their connection). With the current callback-based approach I'll need to start uploads recursively when the previous one finishes, which feels more finicky.

Thanks for this library, by the way -- I'm hoping that once I can get it working the automatic restart/recovery of uploads makes our users' experience much better on unreliable connections!

```java
public class TestUploader {
    public static void upload(final Context context) {
        Log.d("upload", "Going to start the upload on a new background thread");
        Thread backgroundThread = new Thread(new Runnable() {
            @Override
            public void run() {
                // This line throws
                // "java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare()"
                TransferUtility manager = new TransferUtility(
                                            new AmazonS3Client(
                                                    new BasicAWSCredentials("id", "key")),
                                                    context);

                TransferObserver observer = manager.upload("my_bucket", "my_file_name", new File("my_path"));
                observer.setTransferListener(new TransferListener() {
                    @Override
                    public void onStateChanged(int id, TransferState state) {
                        Log.d("upload", "state changed");
                    }

                    @Override
                    public void onProgressChanged(int id, long bytesCurrent, long bytesTotal) {
                        Log.d("upload", "progress changed");
                    }

                    @Override
                    public void onError(int id, Exception ex) {
                        Log.d("upload", "error");
                    }
                });
            }
        });
        backgroundThread.start(); // start the thread in the background
    }
}
```

**Comment by fosterzhang** [comment]
@corbt Awesome. I verified the fix works. Thanks for the code. Stay tuned for the upcoming release.

**Comment by fosterzhang** [comment]
The current implementation can withstand certain connection problem. If network goes off completely, then it will resume when the network comes back. However if the network is crappy, then it may have some trouble. We will improve it's ability to deal with different network condition.

Thanks for your feedback. We will evaluate it and have it prioritized if we decide to implement it.

**Comment by sergiopantoja** [comment]
Thanks for the fix! 

> The current implementation can withstand certain connection problem. ... However if the network is crappy, then it may have some trouble.

@fosterzhang To clarify, is this for the implementation of the worker thread fix, or the implementation of the TransferUtility in general?

**Comment by fosterzhang** [comment]
@sergiopantoja not sure if I understand your question. The bug existed in the implementation of TransferUtility. v2.2.5 released on 08/07 fixes the bug, so using TransferUtility on a worker thread works as expected.

TransferUtility receives connectivity change broadcast. When there is not connectivity, it will pause all transfers and will resume when it comes back. However if the connection is crappy, say times out occasionally, then the TransferUtility will retry several times before it marks the transfer as failure.

**Comment by sergiopantoja** [comment]
@fosterzhang Thank you! That clarifies it for me. Great work.

**Comment by fosterzhang** [comment]
Closing this issue, as it was fixed in [release v2.2.5](https://github.com/aws/aws-sdk-android/releases/tag/release_v2.2.5). Please check out the [release notes](http://aws.amazon.com/releasenotes/Android/3135174622437457) for more details.
